File: 灵茶/图论/拓扑排序/0207.java
import java.util.*;

class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] indegrees = new int[numCourses];
        List<List<Integer>> adjacency = new ArrayList<>();
        Queue<Integer> queue = new LinkedList<>();

        for (int i = 0; i < numCourses; i++) {
            adjacency.add(new ArrayList<>());
        }

        for (int[] pre : prerequisites) {
            indegrees[pre[0]]++;
            adjacency.get(pre[1]).add(pre[0]);
        }

        for (int i = 0; i < numCourses; i++) {
            if (indegrees[i] == 0) {
                queue.add(i);
            }
        }
        
        while (!queue.isEmpty()) {
            int item = queue.poll();
            numCourses--;
            for (int nxt : adjacency.get(item)) {
                indegrees[nxt]--;
                if (indegrees[nxt] == 0) {
                    queue.add(nxt);
                }
            }
        }
        return numCourses == 0;
    }

}

File: 灵茶/图论/拓扑排序/0207.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        indegrees = [0 for _ in range(numCourses)]
        adjacency = [[] for _ in range(numCourses)]
        queue = []
        index = 0
        for i, j in prerequisites:
            indegrees[i] += 1
            adjacency[j].append(i)

        for i in range(len(indegrees)):
            if indegrees[i] == 0:
                queue.insert(0, indegrees[i])
        
        while queue:
            node = queue.pop()
            index += 1
            for nxt in adjacency[node]:
                indegrees[nxt] -= 1
                if indegrees[nxt] == 0:
                    queue.insert(0, nxt)
        
        return index == numCourses
        
